### PR TITLE
gh-107845: Fix symlink handling for tarfile.data_filter

### PR DESCRIPTION
**Pull request by encukou**
* Issue: gh-107845

----
:books: Documentation preview :books:: https://cpython-previews--107846.org.readthedocs.build/

### CODE REVIEW

**Review by ethanfurman — APPROVED** [review]

File: Lib/test/test_tarfile.py [hunk 1]
```diff
@@ -3491,8 +3496,15 @@ def test_parent_symlink(self):
         # Test interplaying symlinks
         # Inspired by 'dirsymlink2a' in jwilk/traversal-archives
         with ArchiveMaker() as arc:
+
+            # links to `.` which is both:
```
Comment (frenzymadness) [inline Lib/test/test_tarfile.py]:
```suggestion
            # `current` links to `.` which is both:
```

**Review by frenzymadness — COMMENTED** [review]
I've found only one comment which might be improved.

As I'm thinking about more possible tests, I can imagine two more types:
1. Chains of symlinks/hardlinks, something like `symlink → hardlink → file` or `hardlink → symlink → file` etc.
2. (This can be combined with idea 1) Tests of archives where the symlinks/hardlinks are added before their destinations. So for the first example above, it'd be add symlink pointing to nonexisting hardlink, add hardlink to nonexisting file, and then add the target file.

File: Doc/library/tarfile.rst [hunk 2]
```diff
@@ -740,6 +740,11 @@ A ``TarInfo`` object has the following public data attributes:
    Name of the target file name, which is only present in :class:`TarInfo` objects
    of type :const:`LNKTYPE` and :const:`SYMTYPE`.
 
+   For symbolic links (``SYMTYPE``), the linkname is relative to the directory
```
Comment (vstinner) [inline Doc/library/tarfile.rst]:
```suggestion
   For symbolic links (``SYMTYPE``), the *linkname* is relative to the directory
```

File: Doc/library/tarfile.rst [hunk 3]
```diff
@@ -740,6 +740,11 @@ A ``TarInfo`` object has the following public data attributes:
...
+   that contains the link.
+   For hard links (``LNKTYPE``), the linkname is relative to the root of
```
Comment (vstinner) [inline Doc/library/tarfile.rst]:
```suggestion
   For hard links (``LNKTYPE``), the *linkname* is relative to the root of
```

**Review by vstinner — APPROVED** [review]
LGTM.

The doc and tarfile file LGTM. I didn't have time to review carefully the tests. Maybe first merge into main, see how it goes on buildbots, and only later backport to all branches. Sometimes, we have bad surprises.

File: Lib/test/test_tarfile.py [hunk 4]
```diff
@@ -3593,12 +3657,30 @@ def test_absolute_symlink(self):
         with self.check_context(arc.open(), 'data'):
             self.expect_exception(
                 tarfile.AbsoluteLinkError,
-                "'parent' is a symlink to an absolute path")
+                "'parent' is a link to an absolute path")
+
+    def test_absolute_hardlink(self):
+        # Test hardlink to an absolute path
+        # Inspired by 'dirsymlink' in jwilk/traversal-archives
```
Comment (vstinner) [inline Lib/test/test_tarfile.py]: Are you talking about https://github.com/jwilk/traversal-archives ? If yes, I suggest writing the whole URL:

```suggestion
        # Inspired by 'dirsymlink' in https://github.com/jwilk/traversal-archives
```

**Comment by encukou** [comment]
AFAIK you can't extract a hardlink to file that's later in the archive (unless the file already exists on disk). Tarfile only searches entries before the link.
I added a test for link chains.

**Comment by encukou** [comment]
Can't seem to get the buildbots to run, so I'll watch them after merging.

**Pull request merged**